src/box/memtx_space.c

```diff
@@ -281,33 +281,33 @@ memtx_space_replace_all_keys(struct space *space, struct tuple *old_tuple,
 	 * Since modification of ephemeral spaces are allowed without txn,
 	 * we must not use MVCC for those spaces even if txn is present now.
 	 */
-	if (memtx_tx_manager_use_mvcc_engine && !space->def->opts.is_ephemeral) {
-		struct txn_stmt *stmt = txn_current_stmt(in_txn());
-		return memtx_tx_history_add_stmt(stmt, old_tuple, new_tuple,
-						 mode, result);
-	}
-
+	bool needs_txm = memtx_tx_manager_use_mvcc_engine &&
+			 !space->def->opts.is_ephemeral;
 	/*
 	 * If old_tuple is not NULL, the index has to
 	 * find and delete it, or return an error.
 	 */
+	int (*index_replace_impl)(struct index *, struct tuple *,
+				  struct tuple *, enum dup_replace_mode,
+				  struct tuple **, struct tuple **) =
+		needs_txm ? index_replace_with_txm : index_replace;
```

Comment:
Dealing with a function pointer looks ugly. I'd make `index_replace_with_txm` decide whether it actually needs to use txm or not:

```C
static int
memtx_index_replace(...)
{
        if (!memtx_tx_manager_use_mvcc_engine)
                return index_replace(...);
        ...
}
```

Reply:
Refactored.